<center><a target="_blank" href="https://academy.constructor.org/"><img src=https://lh3.googleusercontent.com/d/1EmH3Jks5CpJy0zK3JbkvJZkeqWtVcxhB width="800" style="background:none; border:none; box-shadow:none;" /></a> </center>

_____
<center> <h1>  </h1> </center>
<center> <h1> Live coding 4: Functions in Python </h1> </center>

<p style="margin-bottom:1cm;"></p>

_____

<center>Constructor Academy, 2025</center>

<div style="background:#EEEDF5;border-top:0.1cm solid #EF475B;border-bottom:0.1cm solid #EF475B;">
    <div style="margin-left: 0.5cm;margin-top: 0.5cm;margin-bottom: 0.5cm;color:#303030">
        <p><strong>Goal:</strong> Short description of the aim of the live coding session</p>
        <strong> Outline:</strong>
        <a id='P0' name="P0"></a>
        <ol>
            <li> <a style="color:#303030" href='#I'>Introduction </a> </li>
            <li> <a style="color:#303030" href='#P1'>Defining Functions</a></li>
            <li> <a style="color:#303030" href='#P2'>Passing Arguments to Functions</a></li>
            <li> <a style="color:#303030" href='#P3'>Return Values</a></li>
            <li> <a style="color:#303030" href='#P4'>Calling other Functions and Scope of Variables</a></li>
            <li> <a style="color:#303030" href='#P5'>Built-in Methods</a></li>
            <li> <a style="color:#303030" href='#P6'>Functional Programming</a></li>
        </ol>
        <strong>Topics Trained:</strong> Python Functions, arguments, parameters, Variable scope, methods
    </div>
</div>

<nav style="text-align:right"><strong>
        <a style="color:#00BAE5" href="https://constructor.app/learn/courses/f45190af-1e40-4fb3-b0f4-b062878d55a9/view/content/5ff0f127-9f6c-481d-91fb-27af0a765669" title="momentum"> Module 1, The Data Science Toolkit </a>|
        <a style="color:#00BAE5" href="https://constructor.app/learn/courses/f45190af-1e40-4fb3-b0f4-b062878d55a9/view/content/5ff0f127-9f6c-481d-91fb-27af0a765669/b5c79da5-6bce-4ad0-b750-630d45769526">Day 1, Python Fundamentals </a>|
        <a style="color:#00BAE5" href="https://drive.google.com/uc?export=download&id=1p8hdHs1S_CTJ5uUPbGG2NNPv5BtlB8BY" title="momentum"> Live Coding 4, Functions in Python</a>
</strong></nav>

<a id='I' name="I"></a>
## [Introduction](#P0)

Functions are a great way to re-use your code in many different places, without having to type the same code over and over again. Specifically, a function is a unit of code that works on various inputs (which usually are variables), performs certain operations and produces a certain result.

In other words - using functions, you put together a set of instructions and execute that set of instructions in any place you want in your code. The benefit of this is that, when you need to perform a task multiple times in your code, you do not need to write the same logic over and over again. Instead, you call the function to execute that set of instructions. Using functions makes your code easier to write, read, test and fix bugs.

<a id='P1'></a>
## [Defining Functions](#P0)

Enough of this dry theory - let's have a look at a very simple function in Python:

In [ ]:
# Function definition
def say_hello():
    # Function body
    print("Hello everybody")

The example above shows the simplest function written in Python. To define a function, we use the keyword `def`, followed by the function name, followed by a set of parentheses `()`, and ended with a colon `:`. This informs the Python interpreter that we are defining a function.

The following code block, meaning all indented lines that follow the function definition, belongs to that function. The code block is also called the function body.

To use the function, you can just call it by it's name. This will execute the function body that belongs to this function. And you can call it as many times you want:

In [ ]:
say_hello()  # Function call => "Hello everybody"
say_hello()  # Another function call => "Hello everybody"
say_hello()  # And yet another function call => "Hello everybody"

Hello everybody
Hello everybody
Hello everybody


<a id='P2'></a>
## [Passing Arguments to Functions](#P0)

The parenthesis in a function definition are used to define parameters. Parameters can be super useful to create functions that can be re-used for different purposes.

Let's modify our function `say_hello()` in a way, that greets each user individually by name. To do that, we can pass the name of the user as an information to the function as *parameter*.

In [ ]:
def say_hello(name):
    print(f"Hello {name}")

say_hello("Laurent") # => Hello Laurent
say_hello("Adriano")   # => Hello Adriano
say_hello("Class")   # => Hello Class

Hello Laurent
Hello Adriano
Hello Class


The variable `name` in the function definition `say_hello(name)` is an example of a parameter.
Parameters are the main way of passing all necessary information to a function, that it needs to do its job.
The values `'Laurent', 'Simon', 'Class'` are called arguments.
Arguments are that piece of information you pass to a parameter when you call the function.

You can pass arguments to a function in a lot of ways. The most common way is to pass arguments is as *positional arguments*, in which you pass arguments in the same order in which the function parameters were listed.

To make function calls more readable, it sometimes makes sense to list arguments together with the parameter name together. This is also known as *keyword arguments*.

### Positional Arguments

Positional arguments are always provided in the same order as the function parameters have been defined.

For example:

In [ ]:
def describe_movie(title, genre):
    return "{} is a {} movie".format(title, genre)

# correct
print(describe_movie('Star Wars: Episode IV - A New Hope', 'Action, Adventure, Fantasy'))

# arguments order switched around --> wrong result
print(describe_movie('Action, Adventure, Fantasy', 'Star Wars: Episode IV - A New Hope'))

Star Wars: Episode IV - A New Hope is a Action, Adventure, Fantasy movie
Action, Adventure, Fantasy is a Star Wars: Episode IV - A New Hope movie


The definition of `describe_movie` needs a movie title and a genre.
In the function call above, we provide title and genre in the same order as in the function definition.
Obviously, you are going to have unexpected and wrong results if you switch the order of the arguments when the function is being called.

### Keyword Arguments

*Keyword arguments* are name-value pairs that you pass to the function when you call it.
It is a great way of not confusing the order of arguments with the parameters.
It also can make the code easier to understand.

Using keyword_arguments the `describe_movie` function will look like this:

In [ ]:
def describe_movie(title, genre):
    return "{} is a {} movie".format(title, genre)

# correct
print(describe_movie(title='Star Wars: Episode IV - A New Hope',
                     genre='Action, Adventure, Fantasy'))

Star Wars: Episode IV - A New Hope is a Action, Adventure, Fantasy movie


In [ ]:
# argument order switched around -> correct, because we are using keyword arguments
print(describe_movie(genre='Action, Adventure, Fantasy',
                     title='Star Wars: Episode IV - A New Hope'))

Star Wars: Episode IV - A New Hope is a Action, Adventure, Fantasy movie


### Default Values

When you define a function, you can define a default value for one or more of its parameters.
Default values are a great way to reduce the amount of typing, especially when one of the parameters doesn't change that often between different function calls.

For example:

In [ ]:
def describe_movie(title, genre="Fantasy"):
    return "{} is a {} movie".format(title, genre)


# "Star Wars: Episode IV - A New Hope is a Fantasy movie
print(describe_movie(title='Star Wars: Episode IV - A New Hope'))

# "Star Wars: Episode IV - A New Hope is a Science Fiction movie
print(describe_movie(title='Star Wars: Episode IV - A New Hope', genre='Science Fiction'))

Star Wars: Episode IV - A New Hope is a Fantasy movie
Star Wars: Episode IV - A New Hope is a Science Fiction movie


In [ ]:
def add_one_list(mylist=None):
    if mylist:
        mylist.append(1)
        return mylist
    else:
        return [1]

In [ ]:
add_one_list()

[1]

In [ ]:
add_one_list()

[1]

As you can see in the example above, if you pass an argument to a default parameter, the new value will be used instead of the default value.

There are cases when you want to make an argument optional. Default values combined with an if statement are a great way to do that:

In [ ]:
def full_name(first_name, last_name, prefix=None):
    full_name = first_name + " " + last_name

    if prefix:
        full_name = prefix + " " + full_name

    return full_name


print(full_name("Tommy Lee", "Jones"))
print(full_name("Mark", "Hamill", "Mr."))

Tommy Lee Jones
Mr. Mark Hamill


In the example above, `prefix` is optional. It is basically an additional parameter, that can be used if needed, but will be ignored by the function if not used.

### Adding an Arbitrary Number of Arguments

Most often, you already know exactly how many arguments you want to receive, and how those arguments are named.
However, there are some cases, in which you don't know yet if you want to receive one, two, or 100 input values.

Python supports so-called *variable arguments*.
Variables in this context means, that the number of arguments and names of arguments is not know beforehand, and can be defined at run-time.

But why would you need that? Take `sum()` as example function.
Ideally, we want to write a function, that can compute the sum of an arbitrary number of arguments:

In [ ]:
def sum(*args):
    result = 0
    for num in args:
        result = result + num
    return result

In [ ]:
sum()

0

In [ ]:
sum(10)

10

In [ ]:
sum(1,2)

3

In [ ]:
sum(1,2,3)

6

And of course, you can combine arguments and variable arguments - just make sure to have the variable arguments at the end of the argument list. Python matches positional and keyword
arguments first, and then collects any remaining arguments in the final variable parameter.

The following example mixes a positional argument with a variable element.

In [ ]:
def ingredients(name, *args):
    print(f"Ingredients for {name}:")
    for ingredient in args:
        print(f'- {ingredient}')


ingredients('Muffin', '2 cups white flour', '1 egg', '1 tablespoon baking powder', '1/2 teaspoons sugar')

Ingredients for Muffin:
- 2 cups white flour
- 1 egg
- 1 tablespoon baking powder
- 1/2 teaspoons sugar


Note that the asterisk `*` together with the parameter name `args` tells Python to make an
empty tuple called args and pack whatever values it receives into this tuple.
The parameter name could be anything - `*my_var_args_list` or `*ingredient_list` works just as well.
However, `*args` has been accepted as the standard name in the Python community.

There is another variant for variable arguments based on keyword arguments `(name=value, ...)`.
In this case, you can use the double-asterisk `**kwargs` to tell Python to create an empty dictionary, and collect any keyword arguments in there.

This case is particularly useful if you want to pass an arbitrary number of arguments but you don't know ahead of time what information you are going to pass on the function call.
The `**kwargs` allows to accept key-value pairs.

Let us look at the following example:

In [ ]:
def user_profile(name, lastname, **kwargs):
    user = {}
    user['name'] = name
    user['lastname'] = lastname
    print(kwargs)
    for key, value in kwargs.items():
        user[key] = value

    return user

user_profile('Donald', 'Duck')

{}


{'name': 'Donald', 'lastname': 'Duck'}

In [ ]:
print(user_profile('Donald', 'Duck', location='Disney', occupation='actor'))

{'location': 'Disney', 'occupation': 'actor'}
{'name': 'Donald', 'lastname': 'Duck', 'location': 'Disney', 'occupation': 'actor'}


The definition of `user_profile()` expects a first and last name, and
then it allows the user to pass in as many name-value pairs as they want. The
double asterisks before the parameter `**kwargs` cause Python to create
an empty dictionary called `**kwargs` and pack whatever name-value pairs it
receives into this dictionary.

Same as with `*args`, you could give `**kwargs` any name, for example `**profile`.
It's the double-asterisk `**` that tells Python that this is a variable argument.

Putting it all together:

```Python
def my_function(a, b, c=1, *d, **e)

    # a & b => mandatory positional arguments
    # c => optional keyword argument
    # *d => arbitrary number of positional arguments, packed into a tuple
    # **f => arbitrary number of keyword arguments, packed into a dict
```

<a id='P3'></a>
## [Return Values](#P0)

Sometimes, you want functions to give values back, instead of displaying results using :code:`print`.
To do that, you can use the :code:`return` statement. Make sure to save the returned value in a variable, in case you want to access it:

In [ ]:
def even_or_odd(value):
    if value % 2 == 0:
        return "even"
    else:
        return "odd"

In [ ]:
# store return value in variable
a = even_or_odd(1)
print(a)

odd


In [ ]:
# using return value directly
print(even_or_odd(1))

odd


A function always returns the value specified by the `return` statement. In the example above, the function `even_or_odd()` returns a string.
The return value can be anything - a single value or a tuple, list, dict and so on.
When you want to return multiple values at the same time, it's very handy to use a tuple: `return ("Hello", 1)`

A function always returns a value, even if you forget to the `return` statement in the function body.
In such cases, the function implicitly will return `None`.

<a id='P4'></a>
## [Calling other Functions and Scope of Variables](#P0)

As you can see from the previous example, the function body can contain any kind of code. And you even can call other functions from a function. Look at this example:

In [ ]:
PI = 3.14159265358979   # global constant -- only place the value of PI is set

def circle_area(radius):
    # local variable
    area = PI*radius*radius    # use value of global constant PI
    return area

In [ ]:
circle_area(2)

12.56637061435916

In [ ]:
area

NameError: name 'area' is not defined

In [ ]:
def circle_circumference(radius):
    # local variable
    circumference = 2*PI*radius         # use value of global constant PI
    return circumference

def print_circle_info(radius):
    # calling print & circle_area function
    print(f'circle area with radius {radius} = {circle_area(radius)}')
    # calling print & circle_circumference function
    print(f'circumference with radius {radius} = {circle_circumference(radius)}')

print_circle_info(5)

circle area with radius 5 = 78.53981633974475
circumference with radius 5 = 31.4159265358979


In [ ]:
print(PI)

3.14159265358979


In [ ]:
print(circumference)

NameError: name 'circumference' is not defined

In [ ]:
print(area)

NameError: name 'area' is not defined

You can see another important concept in the example above - called *scope*. The scope defines in which place of the program a variable can live. `PI` above is defined on the top most level, so it can be accessed by any other place or function within the same file - for example in function `circle_area`, but also in the print statement at the bottom of the code.

`area` and `circumference` are defined inside the *function body*, and can only be accessed there. Therefore, the last two print statements in the example will fail.

<a id='P5'></a>
## [Built-in Methods](#P0)

Methods are just functions that are attached to a class in Python.

You have probably already used plenty of built-in methods. Some that you might already used are **print()**, **''.join()**, **append()**, **split()**, etc

Here you will find the list of [built-in methods](https://www.programiz.com/python-programming/methods/built-in)

<a id='P6'></a>
## [Functional Programming](#P0)

Functional Programming is a coding style that focuses on defining what to do, instead of performing some action. As the name says, it is based on functions. It has its roots in mathematics - where you often also define functions that take some kind of input and give back a result.

In functional programming, the output of the function ONLY depends on the arguments that are passed. Calling the function f for the same value of x should return the same result f(x) no matter how many times you call it.

Characteristics of functional programming:

- Functions are first class citizens, which means there are no restrictions on their use.
- The functions are pure: they may not have any side effects such as reading/writing to a file, writing to console, networking, ...
- Functional programming has ways and constructs to limit the use of for loops
- It has good support for recursion
- Functional programming takes an "expression-oriented" approach

Python provides several functions to support functional programming, like **map**, **filter**, **reduce** and **lambdas**.

### Functions as first class citizen

In Python functions can be passed as arguments of other functions:

In [ ]:
def mul(n):
    return n ** 2

In [ ]:
def mul2(n):
    return (2 * n) ** 2

In [ ]:
def process_data(data, func):
    return func(data) * 3

In [ ]:
process_data(2, mul)

12

In [ ]:
process_data(2, mul2)

48

you can also assign a variable to a function:

In [ ]:
a = mul

In [ ]:
a(2)

4

Finally, you can write a function that returns a function:

In [ ]:
def my_pow(power):
    def func(n):
        return n ** power
    return func

In [ ]:
mul = my_pow(2)

In [ ]:
type(mul)

function

In [ ]:
mul(2)

4

### Lambda Functions

Lambda function are on the-fly functions that don't have a name (anonymous).

``` Python
lambda params: expr(params)
```

Defined functions vs. lambdas

In [ ]:
def mul(n):
    return n ** 2

In [ ]:
mul(2)

4

In [ ]:
a = lambda n: n ** 2

In [ ]:
a(2)

4

In [ ]:
def process_data(data, func):
    return func(data) * 3

In [ ]:
process_data(2, lambda n: n**2)

12

In [ ]:
process_data(2, lambda n: n**4)

48

### Map

Most of the time we apply operations on lists and sequences. What this does is applies that operation on each element of the list or sequence.

For example we easily can update the elements of a list using a **for** loop:

In [ ]:
def multiplication(n):
    mul_by_2 = []
    for item in n:
        mul_by_2.append(item * 2)
    return mul_by_2

multiplication([1,2,3])

[2, 4, 6]

Functional programming allows to define this much simpler and more concise.

Executing the same block of code as before with the *map* function looks like this:

In [ ]:
def multiplication(n):
    return n * 2

list(map(multiplication, [1,2,3]))

[2, 4, 6]

When calling map, provide a function as the first argument and the data as the second argument: **map(fn, iter)**

Often, you don't want to define a full-blown function. To avoid declaring a function, you can use lambdas. Lambda-Functions are anonymous, on-the-fly, unnamed functions. Take a look at this:

In [ ]:
list(map((lambda x: x * 2), [1,2,3]))

[2, 4, 6]

Obviously, you can go crazy with lambdas. You can even use arrays of functions in lambdas, look at this:

In [ ]:
def times_two(a): return a * 2
def times_three(a): return a * 3

numbers = [1, 2, 3, 4]

both_functions = [times_two, times_three]

for number in numbers:
    print(list(map((lambda a: a(number)), both_functions)))

[2, 3]
[4, 6]
[6, 9]
[8, 12]


In [ ]:
print(list(map(lambda number: list(map((lambda a: a(number)), both_functions)), numbers)))

[[2, 3], [4, 6], [6, 9], [8, 12]]


Let's take another example where we want to find the length of each of the elements from a list:

In [ ]:
countries = ['Switzerland', 'Germany', 'France', 'Austria', 'Italy']
lengths = []
for country in countries:
    lengths.append(len(country))
print(lengths)

[11, 7, 6, 7, 5]


We can use map to achieve the same thing:

In [ ]:
lengths = list(map(len, countries))
print(lengths)

[11, 7, 6, 7, 5]


Map is a built-in function; it's always there and works the same way all the time. It also has some performance benefit because it is usually faster than a manually coded for loop.

### Filter

Filter, like map, calls a function on each of the elements of a list. Unlike map it applies a filter to each element. If the function (filter) returns True, then the element will be in the output list. Otherwise, the value will be discarded.

In [ ]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

def is_multiple_of_three(n):
    return n % 3 == 0


print(list(filter(is_multiple_of_three, numbers)))

[3, 6, 9]


We can combine filters with lambdas too, creating the function on the fly for the first argument in the filter.

In [ ]:
numbers = [-1, 0, 1, 2, 3, 4, 5]
print(list(filter((lambda a: a < 2), numbers)))

[-1, 0, 1]


We can also use filter to find the intersections of two list.

In [ ]:
numbers = [-1, 0, 1, 2, 3, 4, 5]
numbers2 = [0, 4, 5, 12, 6, 7, 8]
print(list(filter(lambda a: a in numbers, numbers2)))

[0, 4, 5]


### Reduce

Reduce reduces a list or an iterable to one single value. In Python 3 it lives in the module functools. So, if we want to use it we have to import it first.

In [ ]:
from functools import reduce

numbers = [1, 2, 3, 4, 5, 6]

print(reduce((lambda a, b: a * b), numbers))

720


Ok, let's take a moment to understand what is happening here. ``reduce`` calls the function multiple times. The first parameter will always hold the current result. In our case, this is a. It will always hold the current state of the multiplication. The second parameter, in our case b, will be the current item in the list.

The call sequence will look like this:

- Call #0: a=1, b=2 -> return 2
- Call #1: a=2, b=3 -> return 6
- Call #2: a=6, b=4 -> return 24
- Call #3: a=24, b=5 -> return 120
- Call #4: a=120, b=6 -> return 720

<div style="border-top:0.1cm solid #EF475B"></div>
    <strong><a href='#Q0'><div style="text-align: right"> <h3>End of this Notebook.</h3></div></a></strong>